### ISSUE TITLE
[5.0] Contextual binding doesn't work with method injection.

### ISSUE THREAD

**Issue by Arhimandril** [laravel/framework#6177]
Right now contextual binding resolution only works at the constructor level but it should also work with method injection. So, for example, if we have:

```php
$app->when('PostsController')
    ->needs('PostRepositoryInterface')
    ->give('EloquentPostRepository');
```

then  `PostsController::__construct(PostRepositoryInterface $posts)` will resolve the dependency correctly but `PostsController::store(PostRepositoryInterface $posts)` will only look for a non-contextual binding and will fail to resolve.

**Comment by robinmitra** [comment]
Why was this silently closed without further discussion?

**Comment by hannesvdvreken** [comment]
Because Graham forgot to mention contextual binding is now available in
illuminate/container 5.0

**Comment by nmabhinandan** [comment]
Thanks for the info @hannesvdvreken

**Comment by robinmitra** [comment]
@hannesvdvreken wow that's great! Thanks

**Comment by chargoy** [comment]
@robinmitra Hi! Is it working? Because I have the same issue.

**Comment by smallhadroncollider** [comment]
I can't get this to work.

The following works (but isn't useful):

```
$this->app->bind("API\Http\Requests\RequestInterface", "API\Http\Requests\SquareRequest");
```

This does not:

```
$this->app->when("API\Http\Controllers\SquaresController")
    ->needs("API\Http\Requests\RequestInterface")
    ->give("API\Http\Requests\SquareRequest");
```

**Comment by refear99** [comment]
@smallhadroncollider @hannesvdvreken  
I have the same issue.
use ```bind()``` works,
but use ```when->needs->give``` not working

**Comment by smallhadroncollider** [comment]
@refear99 I've added a pull request which adds a (currently failing) test for this, so hopefully it might be addressed #8906

**Comment by reyezstef** [comment]
In case of form requests, a workaround is by extending the ValidatesWhenResolved interface with your own, like:

~~~
namespace Authentication\Requests\Contracts;
use Illuminate\Contracts\Validation\ValidatesWhenResolved;

interface Validatable extends ValidatesWhenResolved {}
~~~

And than you can bind to that interface:
~~~
$this->app->bind('Authentication\Requests\Contracts\Validatable',
'Authentication\Requests\Login');
~~~

It's acceptable, since I don't like hard dependencies in the blueprint of my controller.

**Comment by jstonne** [comment]
+1 I would like to see contextual binding - `when->needs->give` working on controllers too. I think it is a valid use case @taylorotwell. Thanks for bringing this up @smallhadroncollider.

**Comment by redgenie** [comment]
+1 Not only in controllers but it would be very useful in some Jobs for example. As jobs rely on the injection in the handle() method. Sometimes, you might want to give different instances of the interface to different jobs.

**Comment by feryardiant** [comment]
+1 I had exact same issue with @robinmitra and would love to see it works

**Comment by henriale** [comment]
+1 such solution would make a cleaner and accurate binding

**Comment by nguyentienlong** [comment]
+1 I had the same demand to use this feature

**Comment by nguyentienlong** [comment]
@reyezstef I found a work around solution for this, hope this help
```
$this->app->addContextualBinding(
    "API\Http\Controllers\SquaresController"
    "API\Http\Requests\RequestInterface"
    "API\Http\Requests\SquareRequest"
);
```

**Comment by eric-famiglietti** [comment]
+1 I would love to use contextual binding in a job's `handle()` method.

**Comment by redgenie** [comment]
@nguyentienlong the solution doesn't seem to be working. At least, not with 5.1.40

**Comment by nguyentienlong** [comment]
@redgenie for 2 months I did not work with laravel, but if you want some real example about contextual binding in laravel ( >5), you can refer this article http://longkyanh.info/blog/2016/04/29/contextual-binding-in-laravel-and-o-in-solid-principles/

**Comment by ko1eda** [comment]
It seems every time i start digging into laravels internals I am disappointed to find the feature does not work as intended or I suppose as I had thought/hoped it would.  I see no mention that this only works in the constructor in the official documentation for 5.6.  :-(

**Comment by reyezstef** [comment]
Your class probably does too much when you need to depend on different dependencies in different methods. We should assume code in classes depend on all dependencies inside a class, else we violate some OO class principles. Only for a few reasons we might need binding on method level, this is why Laravel probably supports normal method binding for classes bound to Laravel. Laravel is great for beginners, I do not agree with every Laravel component and the simplicity as it is described in the docs. If Laravel should make if available, I would no use it, simply because I do not agree with the intention of using it. I think the line should be drawn somewhere.

**Comment by karakanb** [comment]
Is there a way to make this work with 5.7+? The only way to inject is constructor, but a lot of stuff can be injected directly into methods, which in this case breaks the standardization within the controller. An example:
```php
public function store(Request $request, MyFactoryInterface $myFactory, Filesystem $cloud) {
    //
}
```

In this case, if you want to inject a different filesystem to this controller method you need to inject it through constructor, which means now you have two injections, one in the constructor and one in the method, which you can also standardize by moving method-level injection to the constructor but that dependency is not required for all the methods in the controller, which makes it unnecessarily resolved. This could be easier if we could inject the dependency into the controller method contextually.

**Comment by joskfg** [comment]
You can do it easily with `bindMethod`

Example:

**AppServiceProvider.php**
```
\App::bindMethod('App\Jobs\MyJob@handle', function ($job, $app) {
    return $job->handle(
        \App::make(Depdency1::class),
        \App::make(Depdency2::class),
        config('my-config'),
        env('SOMETHING')
    );
}
```

**Comment by 4n70w4** [comment]
+1 for contextual binding in a job's handle() method.

**Comment by d1am0nd** [comment]
Here's a case for it:

```
<?php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Repositories\Contracts\GetAllComedians;
use App\Transformers\Contracts\ComedianTransformer;
use App\Transformers\Contracts\ExtractHitsTransformer;

class ComedianController extends Controller
{
    private $repository;

    private $hitsTransformer;

    private $comedianTransformer;

    public function __construct(
        GetAllComedians $repository,
        ExtractHitsTransformer $hitsTransformer,
        ComedianTransformer $comedianTransformer
    ) {
        $this->repository = $repository;
        $this->hitsTransformer = $hitsTransformer;
        $this->comedianTransformer = $comedianTransformer;
    }

    public function index() {
        return array_map(
            $this->comedianTransformer,
            $this->hitsTransformer->__invoke(
                $this->repository->getAllComedians()
            )
        );
    }
}
```

If I could typehint HitsTransformer in the method it would be much nicer than explicitly calling `->__invoke(..)`

**Comment by rela589n** [comment]
TL; DR.
Laravel now is version 7. Can we bind dependencies into methods like this?
```
$this->app->when(DeleteOverdueFilesJob::class)
            ->method('handle')
            ->needs(FilesRepository::class)
            ->give(EloquentFilesRepository::class);
```

**Comment by driesvints** [comment]
Hey all. I'm going to lock this thread for now as we've had this request numerous time in the past but we decided that we won't be supporting this long ago. Sorry about that.
